RDE: Add rde operations support.

raja-gade opened the pull request:
RDE: Add rde operations support.

Added GET  and PATCH operations on RDE resource URIs
handleProcessorRoot: To display all processor RDE URIs
handleProcRDE: Support GET operations on resources
handleDeferredBindings: Formats the deferred bindings

Commands tested:
'''
// Updated the below to display "oem/AMD/SocConfiguration" 
curl -s -k -u root:0penBmc -H "Content-type: application/json"   -X GET https://<BMC_IP>/redfish/v1/Systems/system/Processors/<ProcessorId>

curl -s -k -u root:0penBmc -H "Content-type: application/json"   -X GET https://<BMC_IP>/redfish/v1/Systems/system/Processors/<ProcessorID>/oem/AMD/SocConfiguration
curl -s -k -u root:0penBmc -H "Content-type: application/json"   -X GET https://<BMC_IP>/redfish/v1/Systems/system/Processors/<ProcessorID>/oem/AMD/SocConfiguration/<path>
curl -s -k -u root:0penBmc -H "Content-type: application/json"   -X PATCH https://<BMC_IP>/redfish/v1/Systems/system/Processors/<ProcessorID>/oem/AMD/SocConfiguration/<path> -d "{xxxx}"

'''

Signed-off-by: Raja Sekhar Reddy Gade <rajagade@amd.com>

raja-gade committed:
RDE:Add RDE base Support for AMD SoC Configuration

Enhanced Redfish endpoint to expose AMD OEM extension:
  /redfish/v1/Systems/<systemId>/Processors/<processorId>/oem/AMD/SocConfiguration

Added support to display AMD SoC Configuration Tokens via RDE.

Parsed rde_device_metadata.json to map processor UUIDs to instances.

Tested:

'''
Eg:
Content of rde_device_metadata.json file
{
  "Processors": {
    "P0": {
      "UUIDs": [
        "xxxx",
        "xxxx"
      ]
    },
    "P1": {
      "UUIDs": [
        "xxxx",
        "xxxx"
      ]
    }
  }
}

Commands tested:
curl -s -k -u root:0penBmc -H "Content-type: application/json" -X GET https://<BMC_IP>/redfish/v1/Systems/<systemId>/Processors/<processorId>

curl -s -k -u root:0penBmc -H "Content-type: application/json"   -X GET https://<BMC_IP>/redfish/v1/Systems/<systemId>/Processors/<processorId>/oem/AMD/SocConfiguration
'''

Signed-off-by: Raja Sekhar Reddy Gade <rajagade@amd.com>

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,22 +18,68 @@
+std::string redfishSystemsBase = "/redfish/v1/Systems/system";
+std::string pldmService = "xyz.openbmc_project.PLDM";
+std::map<uint8_t, std::string> eidUUIDMap;
+std::vector<std::string> uuidMapP0 = {"e5aae21f-b3fa-40a1-918d-f8a19d34ff0e",
```
The metadata should originate from an AMD-installed path. A dedicated JSON file may be created to store relevant metadata such as UUIDs and system base paths. If this file exists, the service should use the custom data defined within it; otherwise, it should fall back to default path resolution.
AMD-specific metadata should not be hardcoded or maintained within BMCWeb itself. 

Example JSON Metadata File (rde_redfish_metadata.json)
{
  "P0": {
    "UUIDs": [
      "xxxxxxx",
      "xxxxxxx"
    ],
    "RedfishRoot": "/redfish/v1/Systems/system/Processor/0"
  },
  "P1": {
    "UUIDs": [
      "xxxxxxx",
      "xxxxxxx"
    ],
    "RedfishRoot": "/redfish/v1/Systems/system/Processor/1"
  }
}

raja-gade replied:
done

aasboddu commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,303 +19,670 @@
-    asyncResp->res.jsonValue = {
-        {"error", "Invalid subUri"}, {"subUri", subUri}, {"status", "failed"}};
-    return;
+    getDevices(req, asyncResp, false);
```
Can you add parse logic in one function. This will help in code reusability.

raja-gade replied:
parse logic involves multiple async D-Bus calls so splitting the API for each call.
We are reusing getDevices() for handleRDE() and handleRDERoot().
 Please let me know your thoughts.

ojayanth replied:
@raja-gade  As discussed, you should optimize the D-Bus interaction by retrieving all the required properties(getall properties) from the device D-Bus in a single call and using them as needed. This approach will help reduce the complexity and overhead of making individual property calls.

raja-gade replied:
done

aasboddu commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,303 +19,670 @@
+        {
+            std::string taskId = task.substr(pos + 1);
+            registerTaskSignal(taskId);
+            taskIDs.push_back(taskId);
```
How are the task IDs obtained? We need to generate them in bmcweb. This will be the operation ID in pldmd for operation commands.

raja-gade replied:
update the code to get different taskIDs for each operation

ojayanth commented on redfish-core/lib/processor.hpp:
```diff
@@ -1323,6 +1323,14 @@ inline void requestRoutesProcessorCollection(App& app)
+        asyncResp->res.jsonValue["AMDSoCConfiguration"]["OEM"] = {
+            {"#Processor.AMDSoCConfiguration",
+             {{"target",
+               "/redfish/v1/Systems/system/Processors/oem/AMDSoCConfiguration/APCB/Tokens"}}},
```
Missing Processor collection id. Please follow the standard Processor schema format. "/redfish/v1/Systems/{}/Processors/{}/

raja-gade replied:
done

ojayanth replied:
Missing Chnages.

ojayanth commented on redfish-core/lib/processor.hpp:
```diff
@@ -1323,6 +1323,14 @@ inline void requestRoutesProcessorCollection(App& app)
+               "/redfish/v1/Systems/system/Processors/oem/AMDSoCConfiguration/APCB/Tokens"}}},
+            {"#Processor.AMDSoCConfiguration.OperationTaskCollection",
+             {{"target",
+               "/redfish/v1/Systems/system/Processors/oem/AMDSoCConfiguration/OperationTasks"}}}};
```
Wy you are Keeping Operationtask under Processor?  /redfish/v1/TaskService/Tasks/{TaskId} is the standard Redfish URI pattern for accessing individual tasks.., Can you provide background on this custom tTaks?

raja-gade replied:
using existing TaskServiceManager . It is removed.

ojayanth replied:
Missing removal ?

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,303 +19,687 @@
-                 "/redfish/v1/Systems/<str>/<str>/<str>/oem/amd/RDE/<path>")
+    BMCWEB_ROUTE(
+        app,
+        "/redfish/v1/Systems/system/<str>/<str>/oem/AMDSoCConfiguration/APCB/Tokens/<path>")
```
Please see comment is processor.hpp , need simliar update here also.

raja-gade replied:
using existing TaskServiceManager . It is removed.

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,303 +19,687 @@
-    BMCWEB_ROUTE(app, "/redfish/v1/Systems/<str>/oem/amd/RDE")
+    BMCWEB_ROUTE(
+        app,
+        "/redfish/v1/Systems/system/<str>/oem/AMDSoCConfiguration/APCB/Tokens")
```
Missing Processor ID here .

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,303 +19,687 @@
         .methods(boost::beast::http::verb::get)(
             std::bind_front(handleRDERoot, std::ref(app)));
+
+    BMCWEB_ROUTE(
```
Please see question in processor.hpp . Please provide clarification why need custom Task URI.

raja-gade replied:
using existing TaskServiceManager . It is removed now

ojayanth commented on redfish-core/lib/processor.hpp:
```diff
@@ -1386,6 +1386,11 @@ inline void requestRoutesProcessor(App& app)
                              "/Actions/Oem/Processor.SupportedErrorTypes"}}}};
 
+        asyncResp->res.jsonValue["OEM"]["AMDSoCConfiguration"] = {
+            {"#Processor.AMDSoCConfiguration",
```
"AmdSocConfigurationToken", ?

raja-gade replied:
done

ojayanth commented on redfish-core/lib/processor.hpp:
```diff
@@ -1386,6 +1386,11 @@ inline void requestRoutesProcessor(App& app)
+        asyncResp->res.jsonValue["OEM"]["AMDSoCConfiguration"] = {
+            {"#Processor.AMDSoCConfiguration",
+             {{"target", "/redfish/v1/Systems/" + systemName + "/Processors/" +
+                             processorId + "oem/AMDSoCConfiguration"}}}};
```
"Oem/AMD/SocConfiguration/  ?

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -131,187 +318,332 @@ inline void handleRDE(crow::App& app, const crow::Request& req,
-                 "/redfish/v1/Systems/<str>/<str>/<str>/oem/amd/RDE/<path>")
+    BMCWEB_ROUTE(
+        app,
+        "/redfish/v1/Systems/<str>/Processors/<str>/oem/AMDSoCConfiguration/<path>")
```
Oem/AMD/SocConfiguration ?

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -131,187 +318,332 @@ inline void handleRDE(crow::App& app, const crow::Request& req,
-    BMCWEB_ROUTE(app, "/redfish/v1/Systems/<str>/oem/amd/RDE")
+    BMCWEB_ROUTE(
+        app,
+        "/redfish/v1/Systems/<str>/Processors/<str>/oem/AMDSoCConfiguration")
```
Oem/AMD/SocConfiguration ?

It would be better to get  supported paths under /redfish/v1/Systems/<str>/Processors/<str>, and extend the uri supported under that.
Like user should see the uri's under https://xxxx/redfish/v1/Systems/system/Processors/<id>

raja-gade replied:
done

ojayanth commented on redfish-core/src/utils/dbus_utils.cpp:
```diff
@@ -18,6 +18,59 @@
 
 namespace redfish
 {
+namespace dbus_utils
```
File scope: Any specific reason for copying the enums from phosphor-dbus-interfaces instead of adding right icludes?

raja-gade replied:
Followed similar to 
https://github.com/AMDESE/bmcweb/blob/integ_sp7/redfish-core/lib/log_services.hpp#L107

Facing some issues for direct includes from phosphor-dbus-interfaces

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -131,187 +318,332 @@ inline void handleRDE(crow::App& app, const crow::Request& req,
+    getDevices(req, asyncResp, processorId, false);
+}
+
+inline void handleRDERoot([[maybe_unused]] crow::App& app,
```
Based on function defintionit is not for generic RDE root , only Processor specific RDE root right? either chnage the function name or make it gerenric to support other schemas.

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -131,187 +318,332 @@ inline void handleRDE(crow::App& app, const crow::Request& req,
+    });
+}
+
+inline void handleRDE(crow::App& app, const crow::Request& req,
```
Based on function definition is not for generic RDE root , only Processor specific RDE root right? either change the function name or make it generic to support other schemas.

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -131,187 +318,332 @@ inline void handleRDE(crow::App& app, const crow::Request& req,
-    asyncResp->res.jsonValue["@odata.id"] = baseUri;
-    asyncResp->res.jsonValue["Name"] = "Redfish Dynamic Extensions";
+    constexpr const char* rdeUUIDCacheFile =
+        "/var/lib/pldm/rde_uuid_cache.json";
```
If the file contains static configuration data that is not expected to change during runtime, consider placing it in /etc/pldm/

ojayanth replied:
Can you example header file now .  to refer the expeceted syntax for uuid config file . you can keep dummy uuid info.

raja-gade replied:
> If the file contains static configuration data that is not expected to change during runtime, consider placing it in /etc/pldm/

done

raja-gade replied:
> Can you example header file now . to refer the expeceted syntax for uuid config file . you can keep dummy uuid info.

included the example in commit message

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -131,187 +318,332 @@ inline void handleRDE(crow::App& app, const crow::Request& req,
+                       const std::string& processorId, bool isRDERoot)
+{
+    reset();
+    loadRDEUUIDMaps();
```
This function skips uuid maps for various error scenario. failure scenario with map initialization , the existing code handles with default or it messes up the the flow?

raja-gade replied:
we are returning with no resource found error.

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -131,187 +318,332 @@ inline void handleRDE(crow::App& app, const crow::Request& req,
+    resourceVect.clear();
+}
+
+inline void getDevices(const crow::Request& req,
```
This function is processor specific , either chnage the function name or make it generic.

raja-gade replied:
done

raja-gade committed:
RDE: Add rde GET operations support.

Added GET operation on RDE resource URIs

handleProcRDE(): Support GET operations on resources

Tested:
Unit tested the code.

Signed-off-by: Raja Sekhar Reddy Gade <rajagade@amd.com>

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,8 +19,34 @@
 
-constexpr std::string_view redfishSystemsBase = "/redfish/v1/Systems";
-constexpr std::string_view oemAmdRDEUri = "/oem/amd/RDE";
+std::string rdeRoot = "/redfish/v1/Systems/system";
```
constexpr?

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,8 +19,34 @@
-constexpr std::string_view redfishSystemsBase = "/redfish/v1/Systems";
-constexpr std::string_view oemAmdRDEUri = "/oem/amd/RDE";
+std::string rdeRoot = "/redfish/v1/Systems/system";
+std::string pldmService = "xyz.openbmc_project.PLDM";
```
constexpr

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
+    BMCWEB_ROUTE(
+        app,
+        "/redfish/v1/Systems/<str>/Processors/<str>/oem/AMD/SocConfiguration/<path>")
         .privileges(redfish::privileges::privilegeSetConfigureManager)
```
Change processor specific privilege's.

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
+    BMCWEB_ROUTE(
+        app,
+        "/redfish/v1/Systems/<str>/Processors/<str>/oem/AMD/SocConfiguration")
         .privileges(redfish::privileges::privilegeSetConfigureManager)
```
Change processor specific privilege's.

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
+        return;
+    }
+
+    getProcessorDevices(req, asyncResp, deviceId, true);
```
getProcessorDevices get all the available Processor RDE devices right? should we limited to Processor ID specific?

raja-gade replied:
For now concentrating for processor. I will make the name generic while enabling it for generic devices.

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
+    resourceVect.clear();
+}
+
+inline void
```
Nit: getProcRdeDevices may be better name

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
+inline void
+    getProcessorDevices(const crow::Request& req,
+                        const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                        const std::string& resourceId, bool isRDERoot)
```
is it resourceId ? based on your code it is deviceId , getting as part of user requested URL?

raja-gade replied:
changed the name to deviceId

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
+                continue;
+            }
+            BMCWEB_LOG_DEBUG("Found device with ObjectPath {}", path);
+            getRDEResourceData(req, asyncResp, resourceId, path, isRDERoot);
```
deviceId ?

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
+            return false;
+        }
+
+        const auto& processors = jsonData[deviceKey];
```
Nit: Better to make dynamically load json data instead of hard coded approch some thing like 
This will help to avoid processor specific array declaration initially . 

for (const auto& [deviceType, deviceData] : jsonData.items())
        {
            if (!deviceData.is_object())
            {
                BMCWEB_LOG_ERROR("RDE: Invalid structure for device type '{}'", deviceType);
                continue;
            }

            for (const auto& [deviceKey, deviceInfo] : deviceData.items())
            {
                if (!deviceInfo.contains("UUIDs") || !deviceInfo["UUIDs"].is_array())
                {
                    BMCWEB_LOG_ERROR("RDE: Missing or invalid 'UUIDs' for '{}.{}'", deviceType, deviceKey);
                    continue;
                }

                for (const auto& uuid : deviceInfo["UUIDs"])
                {
                    if (!uuid.is_string())
                    {
                        BMCWEB_LOG_ERROR("RDE: Invalid UUID format in '{}.{}'", deviceType, deviceKey);
                        continue;
                    }

                    const std::string uuidStr = uuid.get<std::string>();
                    // Example: store in a map of vectors by deviceType and deviceKey
                    uuidMap[deviceType][deviceKey].push_back(uuidStr);
                }
            }
        }
    }

raja-gade replied:
sure will do this optimization in the next iteration

ojayanth replied:
Please include Ticket number(offline) , which captures all the recomenetd chnages , incase not inlcuded as part of this commit.

raja-gade replied:
created a ticket for it

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
-        if (it != expectedSubUris.end())
+
+        std::string deviceID = getDeviceID(*uuid);
+        if (expDeviceID != deviceID)
```
Incase of Mutiple RDE devices in teh system, this is not really error right? caller should take decison on error hanlding instead adding error here ?

raja-gade replied:
updated the code

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
-                        schema["MajorSchemaVersion"].get<std::string>();
+                    std::string uri = normalizePath(*subUri);
+                    res.subUri = uri;
+                    if (!deviceID.empty())
```
Incase of Processor Root empty case is error . start with info  log now.

raja-gade replied:
added info log when deviceID is empty

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +79,380 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
-                         e.what());
-        return schemaMap; // Return empty object
-    }
+std::string normalizePath(const std::string& inputPath)
```
Cleaner way is maintain a map based design  "ProposedContainingResourceName : Uri expected name 
like Processor.Processor->processor
in futute similiar for neteworking schema etc.

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,8 +19,34 @@
+    std::string uuid;
+    std::string schemaName;
+    std::string schemaVersion;
+    int64_t schemaClass;
```
schemaClass PLDM data type(enum8) not int64_t https://github.com/AMDESE/libpldm/blob/df4df207523175b4c2c88033de48afb199adda60/include/libpldm/rde.h#L159 In the d-bus this is string type  not int64_t

raja-gade replied:
We are getting the schemaClass from SchemaResources property, there it is exposed as int64_t
"schemaClass" x 0 

I will change this once it is updated in the SchemaResources  property of pldm

ojayanth replied:
Ack.

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -18,8 +19,34 @@
+std::string pldmService = "xyz.openbmc_project.PLDM";
+std::map<uint8_t, std::string> eidUUIDMap;
+
+std::vector<std::string> uuidMapP0;
```
NIT: In the dymaic design approch , need chnages gere.

raja-gade replied:
sure will do the optimization in the next iteration

ojayanth commented on redfish-core/lib/processor.hpp:
```diff
@@ -1376,6 +1375,11 @@ inline void requestRoutesProcessor(App& app)
             boost::urls::format("/redfish/v1/Systems/{}/Processors/{}",
                                 BMCWEB_REDFISH_SYSTEM_URI_NAME, processorId);
 
+        asyncResp->res.jsonValue["OEM"]["AmdSocConfigurationToken"] = {
```
This will only add URI link , whetre this is going to display supported URI's in redfish/v1?

raja-gade replied:
Yes, this is only for display. Supported resources will be displayed when we do GET on "/redfish/v1/Systems/systemName/Processors/Processor_Id/oem/AMD/SocConfiguration"

ojayanth replied:
Please look https://xxxx/redfish/v1/Managers/bmc/EthernetInterfaces, you can see user can browse here using (@odata.id . you can hook the get backend this .

raja-gade replied:
done

ojayanth commented on redfish-core/lib/processor.hpp:
```diff
@@ -207,7 +207,7 @@ inline void getCpuDataByInterface(
                 }
```
Updated required in the commit message . Not really opertion support . please provide based on the changes.

raja-gade replied:
updated commit message

ojayanth replied:
Still seeing "RDE: Add rde operations support." is this enabled operational support ? Not matching with descriptions. 

Better to split the description to two lines // Updated "/redfish/v1/Systems/<systemId>/Processors/<processorId>" to display "/oem/AMD/SocConfiguration"
curl -s -k -u root:0penBmc -H "Content-type: application/json" -X GET https://<BMC_IP>/redfish/v1/Systems/<systemId>/Processors/<processorId>

curl -s -k -u root:0penBmc -H "Content-type: application/json"   -X GET https://<BMC_IP>/redfish/v1/Systems/<systemId>/Processors/<processorId>/oem/AMD/SocConfiguration

raja-gade replied:
updated the commit message

ojayanth commented on redfish-core/include/utils/dbus_utils.hpp:
```diff
@@ -17,6 +17,34 @@ namespace redfish
 {
 namespace dbus_utils
 {
+enum class OperationType
```
Try using this format in rde.hpp and see you can remove these duplicate definitions.

std::string operationType =
        "xyz.openbmc_project.RDE.Common.OperationType.READ";
    std::string payloadFormat =
        "xyz.openbmc_project.RDE.Manager.PayloadFormatType.Inline";
    std::string encodingFormat =
        "xyz.openbmc_project.RDE.Manager.EncodingFormatType.JSON";

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -48,32 +50,191 @@ struct Resource
-// GET handler
-inline void rdeGetHandler(App& app, const crow::Request& req,
-                          const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
+nlohmann::json handleDeferredBindings(const std::string& bejJsonInput)
```
Better to keep this handleDeferredBindings  as sperate commit

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -107,10 +268,55 @@ std::string normalizePath(const std::string& inputPath)
+        [&expectedUri](const Resource& p) { return p.fullUri == expectedUri; });
+    if (it == resourceVect.end())
+    {
+        BMCWEB_LOG_ERROR("Invalid or missing 'Resources' in schemaMap");
```
Unsupporetd URI error ?

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -48,32 +50,191 @@ struct Resource
+    task->payload.emplace(std::move(payload));
+}
+
+inline uint32_t getOperationID()
```
This is unique from process point. incase bmcweb restarts with some erason , agin reset to 1 . and if we use same operation ID RDE device fails. Need toi add TODO to handle in better way , like persistent approch or something else.

raja-gade replied:
done

raja-gade committed:
RDE: Add API for handleDeferredBindings

handleDeferredBindings: Formats the deferred bindings

Tested:
Able to see the proper data for the defered bingings

Signed-off-by: Aashray Boddu <Aashray.Boddu@amd.com>

raja-gade committed:
RDE: Add rde PATCH operation support.

Added PATCH operation on RDE resource URIs

Tested:
Unit tested the code.

Signed-off-by: Raja Sekhar Reddy Gade <rajagade@amd.com>

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -52,269 +81,386 @@ inline void rdePatchHandler(App& app, const crow::Request& req,
+        if (expDeviceID != deviceID)
         {
-            expectedSubUri = *it;
+            BMCWEB_LOG_INFO(
```
Convert to debug log.

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -53,8 +53,77 @@ struct Resource
 
 std::string handleDeferredBindings(const std::string& bejJsonInput)
 {
```
This commit is from @aasboddu  ? please inlcude him in signed-off.

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -235,14 +235,52 @@ inline void rdeGetHandler(const crow::Request& req,
+    std::string rdePayload = inputJson.dump();
+
+    uint32_t operationID = getOperationID();
+    std::string operationType =
```
constexpr type

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -235,14 +235,52 @@ inline void rdeGetHandler(const crow::Request& req,
+    uint32_t operationID = getOperationID();
+    std::string operationType =
+        "xyz.openbmc_project.RDE.Common.OperationType.UPDATE";
+    std::string payloadFormat =
```
constexpr type

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -235,14 +235,52 @@ inline void rdeGetHandler(const crow::Request& req,
+    std::string operationType =
+        "xyz.openbmc_project.RDE.Common.OperationType.UPDATE";
+    std::string payloadFormat =
+        "xyz.openbmc_project.RDE.Manager.PayloadFormatType.Inline";
```
*constexpr type

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -235,14 +235,52 @@ inline void rdeGetHandler(const crow::Request& req,
+        "xyz.openbmc_project.RDE.Common.OperationType.UPDATE";
+    std::string payloadFormat =
+        "xyz.openbmc_project.RDE.Manager.PayloadFormatType.Inline";
+    std::string encodingType =
```
constexpr type

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -235,14 +235,52 @@ inline void rdeGetHandler(const crow::Request& req,
+        "xyz.openbmc_project.RDE.Manager.PayloadFormatType.Inline";
+    std::string encodingType =
+        "xyz.openbmc_project.RDE.Manager.EncodingFormatType.JSON";
+    std::string sessionID = "1";
```
"sessionID = "1" is hardcoded. ADD todo to get  session tracking  retrieving it dynamically.

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -235,14 +235,52 @@ inline void rdeGetHandler(const crow::Request& req,
+            return;
+        }
+        task::Payload payload(req);
+        rdeCreateTask(asyncResp, std::move(payload), objPath);
```
If rdeCreateTask fails internally, ensure it propagates meaningful errors to the client.

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -50,32 +51,126 @@ struct Resource
+                          std::string targetUri, uint8_t eid, std::string uuid)
+{
+    uint32_t operationID = getOperationID();
+    std::string operationType =
```
constexpr?

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -50,32 +51,126 @@ struct Resource
+    uint32_t operationID = getOperationID();
+    std::string operationType =
+        "xyz.openbmc_project.RDE.Common.OperationType.READ";
+    std::string rdePayload = "";
```
constexpr?

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -50,32 +51,126 @@ struct Resource
+    std::string operationType =
+        "xyz.openbmc_project.RDE.Common.OperationType.READ";
+    std::string rdePayload = "";
+    std::string payloadFormat =
```
constexpr?

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -50,32 +51,126 @@ struct Resource
+    std::string rdePayload = "";
+    std::string payloadFormat =
+        "xyz.openbmc_project.RDE.Manager.PayloadFormatType.Inline";
+    std::string encodingType =
```
constexpr?

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -50,32 +51,126 @@ struct Resource
+        "xyz.openbmc_project.RDE.Manager.PayloadFormatType.Inline";
+    std::string encodingType =
+        "xyz.openbmc_project.RDE.Manager.EncodingFormatType.JSON";
+    std::string sessionID = "1";
```
Add TODo to get this dynamically.

raja-gade replied:
done

ojayanth commented on redfish-core/lib/rde.hpp:
```diff
@@ -50,32 +51,126 @@ struct Resource
+            return;
+        }
+        task::Payload payload(req);
+        rdeCreateTask(asyncResp, std::move(payload), objPath);
```
Add error handling if the task failed inetrnally.

raja-gade replied:
task manager is taking care of the errors

raja-gade commented:
Closing this PR.
https://github.com/AMDESE/bmcweb/pull/31 have the changes
